google/resource_storage_object_iam_policy.go

```diff
@@ -0,0 +1,115 @@
+package google
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	storagev1 "google.golang.org/api/storage/v1"
+)
+
+func resourceStorageObjectIAMPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceStorageObjectIAMPolicyCreate,
+		Read:   resourceStorageObjectIAMPolicyRead,
+		Update: resourceStorageObjectIAMPolicyUpdate,
+		Delete: resourceStorageObjectIAMPolicyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"object": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"policy_data": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: jsonPolicyDiffSuppress,
+				ValidateFunc:     validateV2IamPolicy,
+			},
+			"etag": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func setObjectIamPolicy(d *schema.ResourceData, config *Config) error {
+	bucket := d.Get("bucket").(string)
+	object := d.Get("object").(string)
+	policy, err := unmarshalStorageIamPolicy(d.Get("policy_data").(string))
+	if err != nil {
+		return fmt.Errorf("'policy_data' is not valid for %s: %s", object, err)
+	}
+
+	log.Printf("[DEBUG] Setting IAM policy for object %q on bucket %q", object, bucket)
+	_, err = config.clientStorage.Objects.SetIamPolicy(bucket, object, policy).Do()
+	log.Printf("[DEBUG] Set IAM policy for object %q on %q", object, bucket)
+	return err
+}
+
+func resourceStorageObjectIAMPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	if err := setObjectIamPolicy(d, config); err != nil {
+		return err
+	}
+
+	d.SetId(d.Get("bucket").(string) + "-" + d.Get("object").(string) + "-iam-policy")
```

Comment:
If we could use a separator here that is an invalid character for a bucket name, that'd be awesome. (`/` is a good choice, I believe.)

The idea is that we can then parse the ID to get the bucket and object back, which helps in a few places. As it stands, there's no real way to reliably turn this ID back into its constituent parts. In `my-bucket-object-id-iam-policy`, for example, we'd have no idea what was the bucket and what was the object.